File: restaurent/lib/model/food.dart
import 'package:restaurent/model/Carousel.dart';
import 'package:restaurent/model/Category.dart';
import 'package:restaurent/model/NutritionValue.dart';
import 'package:restaurent/model/Comment.dart';


class Food {
  final int id;
  final String name;
  final String image;
  final String description;
  final String link ;
  final double rating;
  final double price;
  final Category category;
  final NutritionValue nutritionValue;
  final int totalCalories;
  final List<Comment> comments;
  final Carousel carousel;

  Food({
    required this.id,
    required this.name,
    required this.image,
    required this.description,
    required this.rating,
    required this.price,
    required this.link,
    required this.category,
    required this.nutritionValue,
    required this.totalCalories,
    required this.carousel,
    required this.comments,
    
  });

  factory Food.fromJson(Map<String, dynamic>? json) {
    if (json == null) {
      throw ArgumentError.notNull('json');
    }

    // * Parse comments
    List<Comment> parsedComments = [];
    if (json['comments'] != null) {
      var commentsList = json['comments'] as List<dynamic>;
      parsedComments =
          commentsList.map((comment) => Comment.fromJson(comment)).toList();
    }


    return Food(
      id: json['id'] ?? 0,
      name: json['name'] ?? '',
      image: json['image'] ?? '',
      link: json['link'] ?? '',
      description: json['description'] ?? '',
      rating: (json['rating'] ?? 0.0).toDouble(),
      price: (json['price'] ?? 0.0).toDouble(),
      category: Category.fromJson(json['category'] ?? {}),
      nutritionValue: NutritionValue.fromJson(json['nutritionValue'] ?? {}),
      totalCalories: json['totalCalories'] ?? 0,
      carousel: Carousel.fromJson(json['carousel'] ?? {}),
      comments: parsedComments,
      
    );
  }
  
}

File: MyBackEnd/src/main/java/com/example/mybackend/Models/Food.java
package com.example.mybackend.Models;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.Comments;

import java.util.List;
import java.util.Map;


@Entity
@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class Food {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long id;

    private String name;
    private String image;
    private String link;

    @Lob
    @Column(length = 10000)  // * Specifies the maximum length of the description
    private String description;

    private float rating;
    private float price;

    @ManyToOne
    @JoinColumn(name = "category_id")
    @JsonManagedReference
    private Category category;

    @ElementCollection
    @CollectionTable(name = "food_nutrition", joinColumns = @JoinColumn(name = "food_id"))
    @MapKeyColumn(name = "nutrition_name")
    @Column(name = "nutrition_value")
    private Map<String, Float> nutritionValue;

    private int totalCalories;

    @OneToMany(mappedBy = "food", cascade = CascadeType.ALL)
    @JsonManagedReference
    private List<Comment> comments;

    @OneToOne(mappedBy = "food", cascade = CascadeType.ALL)
    private Carousel carousel;

    @ManyToMany(mappedBy = "favoriteFoods")
    @JsonIgnore
    private List<User> users;

    @OneToMany(mappedBy = "food")
    @JsonManagedReference
    @JsonIgnore
    private List<Orders> orders;

    @Override
    public String toString() {
        return "Food{" +
                "id=" + id +
                ", name='" + name + '\'' +
                '}';
    }

}
